V3K molfile: Error while parsing index atom X in bond

Given the following V3K molfile:

```

  Mrv2219 11302223302D          

  0  0  0     0  0            999 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 2 1 0 0 0
M  V30 BEGIN ATOM
M  V30 1 C -6.458 3.4367 0 0
M  V30 22 C -7.7917 2.6667 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 22
M  V30 END BOND
M  V30 END CTAB
M  END
```

I get the following error: "Error while parsing index atom 2 in bond".

AFAICT, the molfile is valid. Replacing atom index 22 in the atom and bond blocks resolves the error.

Here's the stack trace:

```
org.openscience.cdk.io.MDLV3000Reader ERROR: Error while parsing index atom 2 in bond
org.openscience.cdk.exception.CDKException: Error while parsing index atom 2 in bond
        at org.openscience.cdk.io.MDLV3000Reader.readBondBlock(MDLV3000Reader.java:678)
        at org.openscience.cdk.io.MDLV3000Reader.readConnectionTable(MDLV3000Reader.java:193)
        at org.openscience.cdk.io.MDLV3000Reader.readMolecule(MDLV3000Reader.java:166)
        at org.openscience.cdk.io.MDLV3000Reader.read(MDLV3000Reader.java:152)
        at com.companyname.appname.AppTest.testApp(AppTest.java:62)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at junit.framework.TestCase.runTest(TestCase.java:154)
        at junit.framework.TestCase.runBare(TestCase.java:127)
        at junit.framework.TestResult$1.protect(TestResult.java:106)
        at junit.framework.TestResult.runProtected(TestResult.java:124)
        at junit.framework.TestResult.run(TestResult.java:109)
        at junit.framework.TestCase.run(TestCase.java:118)
        at junit.framework.TestSuite.runTest(TestSuite.java:208)
        at junit.framework.TestSuite.run(TestSuite.java:203)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at org.apache.maven.surefire.junit.JUnitTestSet.execute(JUnitTestSet.java:95)
        at org.apache.maven.surefire.junit.JUnit3Provider.executeTestSet(JUnit3Provider.java:121)
        at org.apache.maven.surefire.junit.JUnit3Provider.invoke(JUnit3Provider.java:98)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:577)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
Caused by: java.lang.IndexOutOfBoundsException: No atom at index: 21
        at org.openscience.cdk.silent.AtomContainer2.getAtom(AtomContainer2.java:495)
        at org.openscience.cdk.io.MDLV3000Reader.readBondBlock(MDLV3000Reader.java:672)
        ... 26 more
```

--- a/storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java
+++ b/storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java
@@ -52,6 +52,7 @@ import java.io.InputStreamReader;
 import java.io.Reader;
 import java.io.StringReader;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
 import java.util.Hashtable;
 import java.util.Iterator;
@@ -160,6 +161,38 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
         boolean chiral;
         Map<Integer,Integer> stereoflags = null;
         final Map<IAtom,Integer> stereo0d = new HashMap<>();
+
+        // atom/bond ids need not be sequential, we could use map but more
+        // common is the ids will be sequential
+        IAtom[] atomById = new IAtom[64];
+        IBond[] bondById = new IBond[64];
+
+        <T> T[] grow(T[] arr, int req)
+        {
+            int cap = arr.length;
+            return Arrays.copyOf(arr, Math.max(cap + cap >> 1,
+                                               req + 1));
+        }
+
+        void addAtom(int id, IAtom atom) {
+            if (id >= atomById.length)
+                atomById = grow(atomById, id);
+            atomById[id] = atom;
+        }
+
+        void addBond(int id, IBond bond) {
+            if (id >= bondById.length)
+                bondById = grow(bondById, id);
+            bondById[id] = bond;
+        }
+
+        public IAtom getAtom(int i) {
+            return atomById[i];
+        }
+
+        public IBond getBond(int i) {
+            return bondById[i];
+        }
     }
 
     public IAtomContainer readMolecule(IChemObjectBuilder builder) throws CDKException {
@@ -190,9 +223,9 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
             } else if ("BEGIN ATOM".equals(command)) {
                 readAtomBlock(state);
             } else if ("BEGIN BOND".equals(command)) {
-                readBondBlock(readData);
+                readBondBlock(state);
             } else if ("BEGIN SGROUP".equals(command)) {
-                readSGroup(readData);
+                readSGroup(state);
             } else if ("BEGIN COLLECTION".equals(command)) {
                 readCollection(state);
             } else {
@@ -277,9 +310,9 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                     if (se.getConfigClass() != IStereoElement.TH)
                         continue;
                     IAtom focus = (IAtom) se.getFocus();
-                    int idx = readData.indexOf(focus);
-                    if (idx < 0)
+                    if (focus.getID() == null)
                         continue;
+                    int idx = Integer.parseInt(focus.getID());
                     Integer grpinfo = state.stereoflags.get(idx);
                     if (grpinfo != null)
                         se.setGroupInfo(grpinfo);
@@ -372,7 +405,7 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
             }
             // val-1 since we store atom index instead of atom number
             if (val > 0)
-                flags.put(val-1, type);
+                flags.put(val, type);
             while (i < len && str.charAt(i) == ' ')
                 i++;
             if (i < len && str.charAt(i) == ')')
@@ -455,6 +488,7 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
 
         int RGroupCounter = 1;
         int Rnumber;
+        String id;
         String[] rGroup;
 
         boolean foundEND = false;
@@ -469,7 +503,7 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                 StringTokenizer tokenizer = new StringTokenizer(command);
                 // parse the index
                 try {
-                    atom.setID(tokenizer.nextToken());
+                    id = tokenizer.nextToken();
                 } catch (Exception exception) {
                     String error = "Error while parsing atom index";
                     logger.error(error);
@@ -607,7 +641,9 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                 }
 
                 // store atom
+                atom.setID(id);
                 readData.addAtom(atom);
+                state.addAtom(Integer.parseInt(id), readData.getAtom(readData.getAtomCount()-1));
                 logger.debug("Added atom: " + atom);
             }
         }
@@ -616,7 +652,8 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
     /**
      * Reads the bond atoms, order and stereo configuration.
      */
-    public void readBondBlock(IAtomContainer readData) throws CDKException {
+    public void readBondBlock(ReadState state) throws CDKException {
+        IAtomContainer readData = state.mol;
         logger.info("Reading BOND block");
         boolean foundEND = false;
         while (isReady() && !foundEND) {
@@ -657,7 +694,7 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                 try {
                     String indexAtom1String = tokenizer.nextToken();
                     int indexAtom1 = Integer.parseInt(indexAtom1String);
-                    IAtom atom1 = readData.getAtom(indexAtom1 - 1);
+                    IAtom atom1 = state.getAtom(indexAtom1);
                     bond.setAtom(atom1, 0);
                 } catch (Exception exception) {
                     String error = "Error while parsing index atom 1 in bond";
@@ -669,7 +706,7 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                 try {
                     String indexAtom2String = tokenizer.nextToken();
                     int indexAtom2 = Integer.parseInt(indexAtom2String);
-                    IAtom atom2 = readData.getAtom(indexAtom2 - 1);
+                    IAtom atom2 = state.getAtom(indexAtom2);
                     bond.setAtom(atom2, 1);
                 } catch (Exception exception) {
                     String error = "Error while parsing index atom 2 in bond";
@@ -726,6 +763,8 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
 
                 // storing bond
                 readData.addBond(bond);
+                state.addBond(Integer.parseInt(bond.getID()),
+                              readData.getBond(readData.getBondCount()-1));
 
                 // storing positional variation
                 if ("ANY".equals(attach)) {
@@ -750,7 +789,8 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
     /**
      * Reads labels.
      */
-    public void readSGroup(IAtomContainer readData) throws CDKException {
+    public void readSGroup(ReadState state) throws CDKException {
+        IAtomContainer readData = state.mol;
         boolean foundEND = false;
         while (isReady() && !foundEND) {
             String command = readCommand(readLine());
@@ -788,13 +828,13 @@ public class MDLV3000Reader extends DefaultChemObjectReader {
                                 StringTokenizer atomsTokenizer = new StringTokenizer(value);
                                 int nExpected = Integer.parseInt(atomsTokenizer.nextToken());
                                 while (atomsTokenizer.hasMoreTokens()) {
-                                    sgroup.addAtom(readData.getAtom(Integer.parseInt(atomsTokenizer.nextToken()) - 1));
+                                    sgroup.addAtom(state.getAtom(Integer.parseInt(atomsTokenizer.nextToken())));
                                 }
                             } else if (key.equals("XBONDS")) {
                                 StringTokenizer xbonds = new StringTokenizer(value);
                                 int nExpected = Integer.parseInt(xbonds.nextToken());
                                 while (xbonds.hasMoreTokens()) {
-                                    sgroup.addBond(readData.getBond(Integer.parseInt(xbonds.nextToken()) - 1));
+                                    sgroup.addBond(state.getBond(Integer.parseInt(xbonds.nextToken())));
                                 }
                             } else if (key.equals("LABEL")) {
                                 label = value;

--- a/storage/ctab/src/test/java/org/openscience/cdk/io/MDLV3000ReaderTest.java
+++ b/storage/ctab/src/test/java/org/openscience/cdk/io/MDLV3000ReaderTest.java
@@ -409,6 +409,31 @@ class MDLV3000ReaderTest extends SimpleChemObjectReaderTest {
     }
 
     @Test
+    void testNonSequentialAtomIdx() throws Exception {
+        final String input = "\n" +
+                "  Mrv2219 11302223302D          \n" +
+                "\n" +
+                "  0  0  0     0  0            999 V3000\n" +
+                "M  V30 BEGIN CTAB\n" +
+                "M  V30 COUNTS 2 1 0 0 0\n" +
+                "M  V30 BEGIN ATOM\n" +
+                "M  V30 1 C -6.458 3.4367 0 0\n" +
+                "M  V30 22 C -7.7917 2.6667 0 0\n" +
+                "M  V30 END ATOM\n" +
+                "M  V30 BEGIN BOND\n" +
+                "M  V30 1 1 1 22\n" +
+                "M  V30 END BOND\n" +
+                "M  V30 END CTAB\n" +
+                "M  END";
+        IChemObjectBuilder bldr = SilentChemObjectBuilder.getInstance();
+        try (MDLV3000Reader mdlr = new MDLV3000Reader(new StringReader(input))) {
+            IAtomContainer mol = mdlr.read(bldr.newAtomContainer());
+            Assertions.assertEquals(2, mol.getAtomCount());
+            Assertions.assertEquals(1, mol.getBondCount());
+        }
+    }
+
+    @Test
     void testStereo0d() throws Exception {
         IChemObjectBuilder bldr = SilentChemObjectBuilder.getInstance();
         try (InputStream in = getClass().getResourceAsStream("stereo0d.mdl3");

